File: divoom/snapshot.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// Снапшот старше этого возраста описывает прошлое: Claude Code обновляет его
// только во время сессии.
const staleAfter = 90 * time.Second

type snapshot struct {
	windows map[string]usageWindow
	stale   bool
	// Возраст снапшота; показываем его, когда данные перестали обновляться.
	age time.Duration
	err string
}

type usageWindow struct {
	used        float64
	secondsLeft int
	expired     bool
	present     bool
}

type rawSnapshot struct {
	RateLimits map[string]struct {
		UsedPercentage *float64 `json:"used_percentage"`
		ResetsAt       *float64 `json:"resets_at"`
	} `json:"rate_limits"`
	UpdatedAt string `json:"updated_at"`
}

func snapshotPath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".claude", "usage-snapshot.json"), nil
}

func readSnapshot() snapshot {
	path, err := snapshotPath()
	if err != nil {
		return snapshot{err: err.Error()}
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return snapshot{err: "снапшот не найден"}
	}

	var raw rawSnapshot
	if err := json.Unmarshal(data, &raw); err != nil {
		return snapshot{err: "снапшот повреждён"}
	}

	now := float64(time.Now().UnixNano()) / 1e9
	windows := make(map[string]usageWindow, len(raw.RateLimits))
	for key, entry := range raw.RateLimits {
		if entry.UsedPercentage == nil {
			continue
		}
		w := usageWindow{used: *entry.UsedPercentage, present: true}
		if entry.ResetsAt != nil {
			resetsAt := normalizeEpoch(*entry.ResetsAt)
			if resetsAt <= now {
				// Проценты описывают прошедшее окно, в новом расход
				// начинается с нуля — старое число показывать нельзя.
				w.expired = true
				w.used = 0
			} else {
				w.secondsLeft = int(resetsAt - now)
			}
		}
		windows[key] = w
	}

	if len(windows) == 0 {
		return snapshot{err: "лимитов в снапшоте нет"}
	}

	result := snapshot{windows: windows}
	if updated, err := time.Parse(time.RFC3339, raw.UpdatedAt); err == nil {
		result.age = time.Since(updated)
		result.stale = result.age >= staleAfter
	}
	return result
}

func (s snapshot) window(id string) usageWindow {
	return s.windows[id]
}

// Длина пятичасового окна — по ней считается, сколько его уже прошло.
const fiveHourSeconds = 5 * 60 * 60

func (w usageWindow) fraction() float64 {
	return w.used / 100
}

// elapsedFraction — какая часть окна прожита. Полоса сброса показывает
// именно время, а не расход: в строке статуса это отдельная шкала.
func (w usageWindow) elapsedFraction() float64 {
	if w.expired || w.secondsLeft <= 0 {
		return 1
	}
	return float64(fiveHourSeconds-w.secondsLeft) / fiveHourSeconds
}

func (w usageWindow) percentLabel() string {
	if !w.present {
		return "-"
	}
	return fmt.Sprintf("%.0f%%", w.used)
}

// tint — единственное место с порогами цвета: до 60% базовый цвет окна,
// до 85% оранжевый, дальше красный. Сброшенное окно гасим.
func (w usageWindow) tint(base uint8) uint8 {
	switch {
	case !w.present || w.expired:
		return idxGrey
	case w.used >= 85:
		return idxRed
	case w.used >= 60:
		return idxOrange
	default:
		return base
	}
}

// normalizeEpoch принимает Unix-время в секундах или миллисекундах.
func normalizeEpoch(value float64) float64 {
	if value > 1e12 {
		return value / 1000
	}
	return value
}

File: app/src-tauri/src/snapshot.rs
//! Чтение `~/.claude/usage-snapshot.json` — единственного контракта с writer'ом.
//!
//! Здесь же живут пороги цвета и признак сброшенного окна: фронтенд получает
//! готовый `level` и не дублирует эту логику у себя.

use serde::Serialize;
use std::path::PathBuf;

/// Окна, которые умеет показывать виджет. Добавление нового начинается здесь.
const KNOWN_WINDOWS: &[(&str, &str)] = &[
    ("five_hour", "Сессия · 5 часов"),
    ("seven_day", "Неделя · 7 дней"),
    ("seven_day_opus", "Неделя · Opus"),
];

/// Снапшот старше этого возраста подписывается как устаревший.
const STALE_AFTER_SECONDS: f64 = 90.0;

#[derive(Serialize, Clone, Copy, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Level {
    Ok,
    Warn,
    Critical,
    Expired,
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Window {
    pub id: String,
    pub title: String,
    pub used_percentage: f64,
    /// Секунд до сброса окна; `None`, если сброса нет или он уже наступил.
    pub seconds_left: Option<i64>,
    pub expired: bool,
    pub level: Level,
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Snapshot {
    pub windows: Vec<Window>,
    /// Возраст снапшота в секундах; `None`, если writer не записал `updated_at`.
    pub age_seconds: Option<f64>,
    pub stale: bool,
    /// Текст ошибки вместо данных — писать нечего, показываем подсказку.
    pub error: Option<String>,
}

impl Snapshot {
    pub fn failure(message: &str) -> Self {
        Self {
            windows: Vec::new(),
            age_seconds: None,
            stale: false,
            error: Some(message.to_string()),
        }
    }

    pub fn window(&self, id: &str) -> Option<&Window> {
        self.windows.iter().find(|w| w.id == id)
    }
}

pub fn snapshot_path() -> PathBuf {
    let home = std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_default();
    home.join(".claude").join("usage-snapshot.json")
}

pub fn read() -> Snapshot {
    let path = snapshot_path();

    let Ok(text) = std::fs::read_to_string(&path) else {
        return Snapshot::failure("Запустите сессию Claude Code для получения данных");
    };

    let Ok(root) = serde_json::from_str::<serde_json::Value>(&text) else {
        return Snapshot::failure("Файл со снапшотом повреждён");
    };

    let now = unix_now();
    let limits = root.get("rate_limits");

    let windows: Vec<Window> = KNOWN_WINDOWS
        .iter()
        .filter_map(|(key, title)| {
            let entry = limits?.get(key)?;
            let used = number(entry.get("used_percentage"))?;
            let resets_at = number(entry.get("resets_at")).map(normalize_epoch);

            let expired = resets_at.is_some_and(|at| at <= now);
            let seconds_left = match resets_at {
                Some(at) if !expired => Some((at - now) as i64),
                _ => None,
            };

            Some(Window {
                id: (*key).to_string(),
                title: (*title).to_string(),
                used_percentage: used,
                seconds_left,
                expired,
                level: level_for(used, expired),
            })
        })
        .collect();

    if windows.is_empty() {
        return Snapshot::failure("Запустите сессию Claude Code для получения данных");
    }

    let age_seconds = timestamp(root.get("updated_at")).map(|at| now - at);

    Snapshot {
        stale: age_seconds.is_some_and(|age| age >= STALE_AFTER_SECONDS),
        age_seconds,
        windows,
        error: None,
    }
}

/// Пороги цвета — единственное место на весь проект.
fn level_for(used: f64, expired: bool) -> Level {
    if expired {
        Level::Expired
    } else if used < 60.0 {
        Level::Ok
    } else if used < 85.0 {
        Level::Warn
    } else {
        Level::Critical
    }
}

fn unix_now() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or_default()
}

fn number(value: Option<&serde_json::Value>) -> Option<f64> {
    value?.as_f64()
}

/// Unix-время приходит в секундах или миллисекундах — различаем по величине.
fn normalize_epoch(value: f64) -> f64 {
    if value > 1e12 {
        value / 1000.0
    } else {
        value
    }
}

/// `resets_at` — число, `updated_at` — ISO8601; принимаем оба вида.
fn timestamp(value: Option<&serde_json::Value>) -> Option<f64> {
    let value = value?;

    if let Some(seconds) = value.as_f64() {
        return Some(normalize_epoch(seconds));
    }

    let text = value.as_str()?;
    chrono::DateTime::parse_from_rfc3339(text)
        .ok()
        .map(|dt| dt.timestamp() as f64 + f64::from(dt.timestamp_subsec_millis()) / 1000.0)
}
